libr/core/anal_tp.c

```diff
@@ -71,27 +48,93 @@ static bool type_pos_hit(RAnal *anal, Sdb *trace, bool in_stack, int idx, int si
 		ut64 write_addr = sdb_num_get (trace, sdb_fmt ("%d.mem.write", idx), 0);
 		return (write_addr == sp + size);
 	} else {
-		return sdb_array_contains (trace, sdb_fmt ("%d.reg.write", idx), place, 0);
+		return SDB_CONTAINS (idx, place);
+	}
+}
+
+static void var_rename (RAnal *anal, RAnalVar *v, const char *name, ut64 addr) {
+	if (!name || !v) {
+		return;
+	}
+	bool is_default = strncmp (v->name, "local_", 6)? false: true;
+	if (*name == '*') {
+		name++;
 	}
+	// longer name tends to be meaningful like "src" instead of "s1"
+	if (!is_default && (strlen (v->name) > strlen (name))) {
+		return;
+	}
+	RAnalFunction *fcn = r_anal_get_fcn_in (anal, addr, 0);
+	RAnalVar *v1 = r_anal_var_get_byname (anal, fcn, name);
+	if (v1) {
+		r_anal_var_free (v1);
+		return;
+	}
+	r_anal_var_rename (anal, fcn->addr, 1, v->kind, v->name, name);
 }
 
-static void type_match(RCore *core, ut64 addr, char *name, int prev_idx) {
+static void var_retype (RAnal *anal, RAnalVar *var, const char *vname, char *type, ut64 addr) {
+	if (!type || !var) {
+		return;
+	}
+	if (!strncmp (type, "int", 3)) {
+		return;
+	}
+	char *vtype = NULL;
+	char *ntype = type;
+	if (!strncmp (type, "const ", 6)) {
+		// Droping const from type
+		//TODO: Infering const type
```

Comment:
You might want to check how constant propagation works, could be helpful in this case.

Reply:
yup sure !